File: 프로그래머스/lv2/42885. 구명보트/구명보트.cpp
#include <string>
#include <vector>
#include <deque>
#include <algorithm>
#include <iostream>

using namespace std;

int solution(vector<int> people, int limit) {
    int answer = 0;
    deque<int> dq(people.begin(), people.end());
    sort(dq.begin(), dq.end());
    
    while(dq.size() >= 2) {
        if(dq.front() + dq.back() <= limit) {
            dq.pop_front();
            dq.pop_back();
            answer++;
        } 
        else {
            dq.pop_back();
            answer++;
        }
    }
    if(!dq.empty()) answer++;
    
    return answer;
}

File: 프로그래머스/2/42885. 구명보트/구명보트.py
from collections import deque

def solution(people, limit):
    answer = 0
    
    dq = deque(sorted(people))
    while len(dq) >= 2:
        if dq[0] + dq[-1] <= limit:
            dq.popleft()
            dq.pop()
            answer += 1
        else:
            dq.pop()
            answer += 1
    
    answer = answer + 1 if len(dq) > 0 else answer
    
    return answer